File: code/ex165.py
class Solution:
    def compareVersion(self, version1, version2):
        """
        :type version1: str
        :type version2: str
        :rtype: int
        """

        s1 = version1.split('.')
        s2 = version2.split('.')

        length = min(len(s1), len(s2))

        #Compare the head part
        for i in range(length):
            if int(s1[i]) > int(s2[i]):
                return 1
            
            if int(s1[i]) < int(s2[i]):
                return -1


        #Compare later part of the longer version number
        if len(s1) > len(s2):
            #If all version number in the end is 0.0.0..., two version is the same, else the longer one is larger
            for num in s1[len(s2) : ]:
                if int(num) != 0:   return 1
            return 0
        elif len(s1) < len(s2):
            #If all version number in the end is 0.0.0..., two version is the same, else the longer one is larger
            for num in s2[len(s1) : ]:
                if int(num) != 0:   return -1
            return 0
        else:
            return 0   

sol = Solution()
print(sol.compareVersion('1.1', '1.01.0'))

File: code/ex165.cpp
class Solution {
public:
    int compareVersion(string version1, string version2) {
        vector<int> a1 = split(version1);
        vector<int> a2 = split(version2);
        int n1 = a1.size(), n2 = a2.size(), p1 = 0, p2 = 0;
        while (p1 < n1 && p2 < n2) {
            if (a1[p1] < a2[p2])    return -1;
            else if (a1[p1] > a2[p2])    return 1;
            else {
                p1++;    p2++;
            }
        }
        return n1 > n2 ? 1 : n1 == n2 ? 0 : -1;
    }
    
private:
    vector<int> split(string s1) {
        vector<int> res;
        s1 += '.';
        int temp = 0;
        for (char c : s1) {
            if (c == '.') {
                res.push_back(temp);
                temp = 0;
            } else {
                temp = temp * 10 + (c - '0');
            }
        }
        while (!res.empty() && res.back() == 0)    res.pop_back();
        return res;
    }
};